pkg/apis/condition_set.go

```diff
@@ -0,0 +1,195 @@
+	}
+	return true
+}
+
+// GetCondition finds and returns the Condition that matches the ConditionType
+// previously set on Conditions.
+func (r *ConditionsImpl) GetCondition(t ConditionType) *Condition {
+	if r.conditions == nil {
+		return nil
+	}
+	for _, c := range r.conditions.GetConditions() {
+		if c.Type == t {
+			return &c
+		}
+	}
+	return nil
+}
+
+// SetCondition sets or updates the Condition on Conditions for Condition.Type.
+func (r *ConditionsImpl) SetCondition(new *Condition) {
+	if r.conditions == nil {
+		return
+	}
+	if new == nil {
+		return
+	}
+	t := new.Type
+	var conditions []Condition
+	for _, c := range r.conditions.GetConditions() {
+		if c.Type != t {
+			conditions = append(conditions, c)
+		} else {
+			// If we'd only update the LastTransitionTime, then return.
+			new.LastTransitionTime = c.LastTransitionTime
+			if reflect.DeepEqual(new, &c) {
+				return
+			}
+		}
+	}
+	new.LastTransitionTime = apis.VolatileTime{metav1.NewTime(time.Now())}
+	conditions = append(conditions, *new)
+	sort.Slice(conditions, func(i, j int) bool { return conditions[i].Type < conditions[j].Type })
+	r.conditions.SetConditions(conditions)
+}
+
+// MarkTrue sets the status of t to true, and then marks the lead condition to
+// true if all other dependents are also true.
+func (r *ConditionsImpl) MarkTrue(t ConditionType) {
+	// set the specified condition
+	r.SetCondition(&Condition{
+		Type:   t,
+		Status: corev1.ConditionTrue,
+	})
+
+	// check the dependents.
+	for _, cond := range r.dependents {
+		c := r.GetCondition(cond)
+		// Failed or Unknown conditions trump true conditions
+		if c == nil || c.Status != corev1.ConditionTrue {
+			return
+		}
+	}
+
+	// set the lead condition
+	r.SetCondition(&Condition{
+		Type:   r.lead,
+		Status: corev1.ConditionTrue,
+	})
+}
+
+// MarkUnknown sets the status of t to Unknown and also sets the lead condition
+// to Unknown if no other dependent condition is in an error state.
+func (r *ConditionsImpl) MarkUnknown(t ConditionType, reason, message string) {
+	// set the specified condition
+	r.SetCondition(&Condition{
+		Type:   t,
+		Status: corev1.ConditionUnknown,
+	})
+
+	// check the dependents.
+	for _, cond := range r.dependents {
+		c := r.GetCondition(cond)
+		// Failed conditions trump Unknown conditions
+		if c == nil || c.Status == corev1.ConditionFalse {
+			return
+		}
+	}
+
+	// set the lead condition
+	r.SetCondition(&Condition{
+		Type:    r.lead,
+		Status:  corev1.ConditionUnknown,
+		Reason:  reason,
+		Message: message,
+	})
+}
+
+// MarkFalse sets the status of t and the lead condition to False.
+func (r *ConditionsImpl) MarkFalse(t ConditionType, reason, message string) {
+	for _, t := range []ConditionType{
+		t,
+		r.lead,
+	} {
+		r.SetCondition(&Condition{
+			Type:    t,
+			Status:  corev1.ConditionFalse,
+			Reason:  reason,
+			Message: message,
+		})
+	}
+}
+
+// InitializeConditions updates all Conditions in the ConditionSet to Unknown
+// if not set.
```

Comment:
This appears to set them unconditionally. Intentional?

Reply:
good catch!! no, it is not suppose to do that.